samples/nrf9160/http_application_fota/src/main.c

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <stdint.h>
+#include <zephyr.h>
+#include <flash.h>
+#include <bsd.h>
+#include <nrf_socket.h>
+#include <download_client.h>
+#include <logging/log.h>
+#include <gpio.h>
+#include <flash_map.h>
+#include <dfu/mcuboot.h>
+
+LOG_MODULE_REGISTER(main);
+
+#define LED_PORT	LED0_GPIO_CONTROLLER
+
+static struct	device	*flash_dev;
+static K_THREAD_STACK_DEFINE(dfu_thread_stack, 1024);
+static bool is_flash_page_erased[FLASH_PAGE_MAX_CNT];
+
+static struct	k_thread dfu_thread;
+static bool		start_dfu;
+static struct	gpio_callback gpio_cb;
+
+static int app_download_client_event_handler(struct download_client *const dfu,
+			enum download_client_evt event, u32_t status);
+
+
+static struct download_client dfu = {
+	.host = CONFIG_DOWNLOAD_HOST,
+	.resource = CONFIG_DOWNLOAD_FILE,
+	.callback = app_download_client_event_handler
+};
+
+
+/**@brief Recoverable BSD library error. */
+void bsd_recoverable_error_handler(uint32_t err)
+{
+	while (true) {
+	}
+}
+
+/**@brief Irrecoverable BSD library error. */
+void bsd_irrecoverable_error_handler(uint32_t err)
+{
+	while (true) {
+	}
+}
+
+
+/**@brief Initialize application. */
+static void app_dfu_init(void)
+{
+	int i;
+
+	for (i = 0; i < FLASH_PAGE_MAX_CNT; i++) {
+		is_flash_page_erased[i] = false;
```

Comment:
I'm not sure if the compiler knows to optimize this away since you could do `is_flash_page_erased[FLASH_PAGE_MAX_CNT] = { false };` when you intializes it in the top. This will however limit the capability of calling app_dfu_init multiple times without a reset in between so could be kept as is.